### ISSUE TITLE
FLTK creates C:\FLTK\fltk.org\fltk.prefs

### ISSUE THREAD

**Issue by ricary37** [fltk/fltk#37]
could you please fix your FLTK bug?
https://github.com/TigerVNC/tigervnc/issues/822

**Comment by Albrecht-S** [comment]
Please see my comment to above mentioned TigerVNC/tigervnc#822 issue.
Waiting for further information ...

**Comment by Albrecht-S** [comment]
OK, here's a simple reproducer:
```
#include <FL/Fl.H>
int main() {
  int opt = Fl::option(Fl::OPTION_VISIBLE_FOCUS);
  return opt;
}
```
The effect:
1. if you run this for the first time and `fltk.prefs` does not exist, then it is written.
2. if you run it again and `fltk.prefs` exists then it is not written again.

I assume that (2) was fixed in the above mentioned commits, but (1) is still an issue, even in FLTK 1.4. The issue is, in other words, that **_reading_** FLTK options results in **_writing_** the preferences file. Internal note: see [STR 2823](https://www.fltk.org/str.php?L2823).

Now that we know what's happening we may be able to resolve this.

**Comment by erco77** [comment]
Sounds like this needs an off switch, preferably off by default.
Very unexpected that widgets are triggering anything in the file system by default, no matter the reason.

**Comment by Albrecht-S** [comment]
There's no need for a switch. We "just" ;-) need to fix it so that _reading_ preferences doesn't _write_ the file. Matt tried this already in the commits to STR 2823 but missed the case where the file (or maybe only the particular option) does not exist.

I'm not yet sure how exactly it can be done, but I think we should try it.

**Comment by erco77** [comment]
Here's my concern about widgets interacting with the file system, reading or writing.

My customers often have home directories that are mounted on a network file server, so interacting with widgets that touch the file system would potentially cause network round trips, which is definitely undesirable (subject to unexpected hangs).

And I'm not even sure what the implications would be for embedded devices that might be either read only environments or have limited r/w cycles on the file system.

I'd definitely advocate for either a global switch to turn this off completely, and some docs recommending how to do it for e.g. embedded devices and/or apps that want to operate in very simple environments where touching the file system is bad for numerous reasons (network implications, security perimeters, etc) I know some of my very security conscious customers would care about unexpected file system accesses.

**Comment by Albrecht-S** [comment]
Okay, I see your concerns. But please let's separate this (in theory at least) because we have two different aspects. The one given here is **writing** (even **creating**) a file because we want to **read** an option. We can likely consider this a bug, at least unexpected behavior.

The point you are now mentioning seems to be file system access by widgets (say: the FLTK library) in general, even read-only. Note that Fl::option() is used in many places in the FLTK library but it in general as a read-only feature. If we wanted to avoid file access at all, then we'd need to add a switch that can only be set by calling a function early in the program's initialization, a static global variable, or by an environment variable or something like that. The consequence would be that reading options would always have to fall back to default options or setting options only in memory - which sounds somewhat simple if we'd only disable writing the Fl_Preferences file in the d'tor. Interesting …

However, there'd be user preferences that need to be written, preferences that might be read-only, and other preferences (Fl::options) that might be read-write in memory but not written to the file system.

Something like that, maybe?

**Comment by Albrecht-S** [comment]
I believe this issue is something @MatthiasWM might want to take a look at.

Summary: IMHO the basic question is if `Fl_Preferences` can be **_read_** without **_writing_** any files (as it should be). Otherwise we'd need a more complex solution.

**Comment by MatthiasWM** [comment]
Thanks for including me. The intention was to read only, and if some settings should be set or changed, we would write at some point. I will have to go into the source code that is at least 20 years old to find out what is happening there. I can see that we could have a command that prevents writing or even reading the prefs from ever occurring. I'll try to get at that tonight.

**Comment by Albrecht-S** [comment]
@MatthiasWM: Thanks for your quick reply.

Indeed I can imagine a switch (calling a function) that prevents reading the prefs file (setting all prefs internally to their default values) and also a particular setting that prevents writing to a file even if the prefs were changed.

The latter would enable a program to set, for instance, `Fl::options()` to a particular value w/o needing to access any files. This would work as suggested by @erco77 in the scenario mentioned above in https://github.com/fltk/fltk/issues/37#issuecomment-539684437.

I'm not sure though if this should be (only) a global switch - one should be able to override it per `Fl_Preferences` instance. If we had this, someone could set the global switch to ` 'no read' ` + ` 'no write' ` but still have their own preferences being read and written to files by setting their attributes to read/write.

**Comment by MatthiasWM** [comment]
Let's see, we have system and user preferences, prefs that are read and written by the FLTK core or by the app itself, and the issue that prefs are written, even if prefs were never changed :-) .

RFC: How about this API? The static functions define and query access rights for all instances. The two constructors can be used to override the settings for individual preference files.

```
static void Fl_Preferences::access(USER_READ_OK|SYSTEM_WRITE_OK|CORE_READ_OK);
// plus ALL_OK (default), READ_OK, WRITE_OK, etc.
static int Fl_Preferences::access();
void Fl_Preferences(root, vendor, app, access_flags); // override 'access()'
void Fl_Preferences(path, vendor, app, access_flags); // with READ_OK, WRITE_OK
```

**Comment by erco77** [comment]
For me, a simple 'off' switch for doing anything with files would be sufficient. access(0) sounds fine I guess. The simpler and more intuitive the better.

What you suggest sounds fine for the micro control.

Just make it easy to change just a couple of flags without affecting the others.
I'm guessing with your API a way to do that would be something like:

```
// Remove write for user
int flags = Fl_Preferences::access();
flags &= ~(USER_READ|USER_WRITE);   // just disable r/w for user, leave others alone
Fl_Preferences::access(flags);
```

**Comment by erco77** [comment]
I imagine the names need to be named appropriately for the global namespace, so  either FL_PREFS_USER_READ or Fl_Preferences::USER_READ, or are there maybe better ways? We use both of those in the API currently, /probably/ the latter is better, tho a bit long.

**Comment by Albrecht-S** [comment]
`Fl_Preferences::access()` sounds a little too unspecific, maybe `Fl_Preferences::file_access()` would be more precise.

One condition must be met for a program like TigerVNC (this issue's reason) can work as expected. This code fragment (pseudo code) should disable file access for all FLTK preferences but allow one's own preferences (if any) to be read and written:

```
int main (...) {
  Fl_Preferences::file_access(0);
  Fl::option(...); // reading or setting FLTK options must not read/write pref. files
  // more code here, particularly...
  my_prefs = new Fl_Preferences (..., ENABLE USER READ + WRITE); // own preferences
}
```
Would this work as expected?

**Comment by MatthiasWM** [comment]
I patched Fl_Preferences and wrote docs for the new function and flags. I hope there are no typos. I could not test it on other platforms than macOS. Please let me know what you think, so I can evetually close this issue.

**Comment by Albrecht-S** [comment]
@MatthiasWM - Awesome, thank you very much, this looks great at a first glance:
- builds fine under Linux
- builds fine under Windows (cross-compiled under Linux)
- works well under Linux (as far as I could test right now)

I tested with `strace` under Linux to verify that files were neither read nor written if `Fl_Preferences::file_access()` options have been set accordingly.

@ricary37 - FYI: this has been implemented in FLTK 1.4 (Git commit c0237a1f04) and will probably not be backported to FLTK 1.3. If you or anyone else want to try it: add one of these instructions
```
   Fl_Preferences::file_access(Fl_Preferences::APP_OK); // allow to write own pref. files
   Fl_Preferences::file_access(Fl_Preferences::NONE); // don't allow read/write at all
```
I assume that the 2nd statement would be the right choice for TigerVNC.
